File: src/python/problems/LengthOfLIS.py
#https://leetcode.com/problems/longest-increasing-subsequence/
class LengthOfLIS:
    def Solution(self, nums: list[int]) -> int:
        map = [0] * len(nums)
        for i in range(len(nums) - 1, -1, -1):
            for j in range(i + 1, len(nums)):
                if nums[i] < nums[j]:
                    map[i] = max(map[i], map[j])
            map[i] += 1

        return max(map)
    
    def Solution1(self, nums: list[int]) -> int:
        sub = [nums[0]]
        for i in range(1, len(nums)):
            if nums[i] > sub[-1]:
                sub.append(nums[i])
                continue

            for j in range(len(sub)):
                if nums[i] <= sub[j]:
                    sub[j] = nums[i]
                    break

        return len(sub)


LengthOfLIS().Solution1([4,10,4,3,8,9])

File: src/csharp/Problems/LengthOfLIS.cs
//https://leetcode.com/problems/longest-increasing-subsequence/

namespace LeetCode.Problems;

public sealed class LengthOfLIS : ProblemBase
{
    [Theory]
    [ClassData(typeof(LengthOfLIS))]
    public override void Test(object[] data) => base.Test(data);

    public override void AddTestCases()
        => Add(it => it.ParamArray("[10,9,2,5,3,7,101,18]").Result(4))
          .Add(it => it.ParamArray("[0,1,0,3,2,3]").Result(4))
          .Add(it => it.ParamArray("[0,1,0,4,2,3]").Result(4))
          .Add(it => it.ParamArray("[4,10,4,3,8,9]").Result(3))
          .Add(it => it.ParamArray("[7,7,7,7,7,7,7]").Result(1));

    private int Solution(int[] nums)
    {
        var map = new List<int> { nums[0] };
        for (var i = 1; i < nums.Length; i++)
        {
            if (nums[i] > map[^1])
            {
                map.Add(nums[i]);
                continue;
            }

            for (int j = 0; j < map.Count; j++)
            {
                if (map[j] >= nums[i])
                {
                    map[j] = nums[i];
                    break;
                }
            }
        }

        return map.Count;
    }

    private int Solution1(int[] nums)
    {
        var map = new int[nums.Length];
        for (var i = nums.Length - 1; i >= 0; i--)
        {
            for (int j = i + 1; j < nums.Length; j++)
            {
                if (nums[i] < nums[j])
                {
                    map[i] = Math.Max(map[i], map[j]);
                }
            }

            map[i] += 1;
        }

        return map.Max();
    }
}